File: packages/ios/Sources/QuilttConnector/QuilttConnectorConfiguration.swift
// swift-tools-version: 5.9
// The swift-tools-version declares the minimum version of Swift required to build this package.

import Foundation

public protocol QuilttConnectorConfiguration {
    var connectorId: String { get }
    var appLauncherUrl: String { get }
    var connectionId: String? { get }
    var institution: String? { get }
    var themeMode: String? { get }
}

extension QuilttConnectorConfiguration {
    public var themeMode: String? { nil }
}

public struct QuilttConnectorConnectConfiguration: QuilttConnectorConfiguration {
    public var connectorId: String
    public var appLauncherUrl: String
    public var connectionId: String?
    public var institution: String?
    public var themeMode: String?
    public init(
        connectorId: String,
        appLauncherUrl: String,
        institution: String? = nil,
        themeMode: String? = nil
    ) {
        self.connectorId = connectorId
        self.appLauncherUrl = appLauncherUrl
        self.institution = institution
        self.themeMode = themeMode
    }
}

public struct QuilttConnectorReconnectConfiguration: QuilttConnectorConfiguration {
    public var connectorId: String
    public var appLauncherUrl: String
    public var connectionId: String?
    public var institution: String?
    public var themeMode: String?

    public init(
        connectorId: String,
        appLauncherUrl: String,
        connectionId: String,
        themeMode: String? = nil
    ) {
        self.connectorId = connectorId
        self.appLauncherUrl = appLauncherUrl
        self.connectionId = connectionId
        self.themeMode = themeMode
    }
}

File: packages/android/connector/src/main/java/app/quiltt/connector/QuilttConnectorConfiguration.kt
package app.quiltt.connector

interface QuilttConnectorConfiguration {
    val connectorId: String
    val appLauncherUrl: String
    val connectionId: String?
    val institution: String?
    val themeMode: String?
        get() = null
}

data class QuilttConnectorConnectConfiguration(
    override val connectorId: String,
    override val appLauncherUrl: String,
    override val institution: String? = null,
    override val themeMode: String? = null,
) : QuilttConnectorConfiguration {
    override val connectionId: String? = null // always null for connect, cannot be set
}

data class QuilttConnectorReconnectConfiguration(
    override val connectorId: String,
    override val appLauncherUrl: String,
    override val connectionId: String,
    override val themeMode: String? = null,
) : QuilttConnectorConfiguration {
    override val institution: String? = null // always null for reconnect, cannot be set
}